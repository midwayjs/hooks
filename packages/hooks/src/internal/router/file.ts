import kebabCase from 'lodash/kebabCase'
import last from 'lodash/last'
import { extname, join, relative, removeExt, toUnix } from 'upath'
import { AbstractRouter, Route, urlJoin } from '@midwayjs/hooks-core'
import { camelCase } from 'lodash'

export enum RouteKeyword {
  INDEX = 'index',
  CATCH_ALL = '/*',
  DYNAMIC = ':',
}

export interface FileSystemRouterConfig {
  source: string
  routes: Route[]
  legacy?: boolean
}

type Part = {
  content: string
  dynamic: boolean
  qualifier?: string
  catchAll?: boolean
}

export class FileSystemRouter extends AbstractRouter {
  constructor(public config: FileSystemRouterConfig) {
    super()
  }

  getApiDirectory(baseDir: string) {
    return join(this.config.source, baseDir)
  }

  getFunctionId(file: string, functionName: string, isExportDefault: boolean) {
    const relativePath = relative(this.config.source, file)
    // src/apis/lambda/index.ts -> apis-lambda-index
    const id = kebabCase(removeExt(relativePath, extname(relativePath)))
    const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
    return this.config.legacy ? name.toLowerCase() : camelCase(name)
  }

  getRoute(file: string) {
    return this.config.routes.find((route) => {
      const apiDir = this.getApiDirectory(route.baseDir)
      return this.isPathInside(toUnix(file), toUnix(apiDir))
    })
  }

  isApiFile({ file }) {
    const route = this.getRoute(file)
    return !!route
  }

  functionToHttpPath(
    file: string,
    functionName: string,
    exportDefault: boolean
  ) {
    const { baseDir } = this.getRoute(file)

    const filePath = removeExt(
      relative(this.getApiDirectory(baseDir), file),
      extname(file)
    )

    return urlJoin(this.buildUrl(filePath, exportDefault ? '' : functionName), {
      trailingSlash: false,
    })
  }

  buildUrl(file: string, functionName = '') {
    const parts: Part[] = file
      .split(/\[(.+?\(.+?\)|.+?)\]/)
      .map((str, i) => {
        if (!str) return null
        const dynamic = i % 2 === 1

        const [, content, qualifier] = dynamic
          ? /([^(]+)(\(.+\))?$/.exec(str)
          : [null, str, null]

        return {
          content,
          dynamic,
          catchAll: /^\.{3}.+$/.test(content),
          qualifier,
        }
      })
      .filter(Boolean)

    const catchAllIndex = parts.findIndex((part) => part.catchAll)
    if (catchAllIndex !== -1 && catchAllIndex !== parts.length - 1) {
      throw new Error(
        `Catch all routes must be the last part of the path. Current input: ${file}`
      )
    }

    const segments: string[] = []

    for (let [idx, { content, dynamic, catchAll }] of parts.entries()) {
      if (!content) continue

      if (catchAll) {
        if (content.slice(3) !== RouteKeyword.INDEX) {
          segments.push(content.slice(3))
        }
        continue
      }

      /**
       * /[api]/id -> /:api/id
       */
      if (dynamic) {
        segments.push(`${RouteKeyword.DYNAMIC}${content}`)
        continue
      }

      if (idx === parts.length - 1) {
        const contents = content.split('/')
        if (last(contents) === RouteKeyword.INDEX) {
          contents.pop()
          segments.push(contents.join('/'))
          continue
        }
      }

      segments.push(content)
    }

    segments.push(functionName)
    if (catchAllIndex !== -1) {
      segments.push(RouteKeyword.CATCH_ALL)
    }

    return urlJoin.apply(null, [...segments, {}])
  }
}
