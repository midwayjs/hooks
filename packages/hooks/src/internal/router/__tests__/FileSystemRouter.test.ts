import { FileSystemRouter } from '../file'

describe('NewFileRouter', () => {
  let router: FileSystemRouter

  beforeEach(() => {
    router = new FileSystemRouter({
      source: '/src',
      routes: [
        {
          baseDir: 'lambda',
        },
      ],
    })
  })

  test('should exist', () => {
    expect(FileSystemRouter).toBeTruthy()
    expect(router).toBeInstanceOf(FileSystemRouter)
  })

  test('test file router', () => {
    expect(router.config.source.endsWith('/src')).toBeTruthy()

    const api = '/src/lambda/index.ts'
    expect(router.isApiFile({ file: api })).toBeTruthy()
    expect(router.isApiFile({ file: '/dist' })).toBeFalsy()

    const rule = router.getRoute(api)
    expect(rule).toMatchSnapshot()
    expect(
      router.getApiDirectory(rule.baseDir).endsWith('src/lambda')
    ).toBeTruthy()
  })

  test('file to http path', () => {
    const cases = [
      {
        file: '/src/lambda/index.ts',
        functionName: '',
        exportDefault: true,
        expected: '/',
      },
      {
        file: '/src/lambda/index.ts',
        functionName: 'index',
        exportDefault: true,
        expected: '/',
      },
      {
        file: '/src/lambda/index.ts',
        functionName: '',
        exportDefault: true,
        expected: '/',
      },
      {
        file: '/src/lambda/index.ts',
        functionName: 'index',
        exportDefault: false,
        expected: '/index',
      },
      {
        file: '/src/lambda/foo.ts',
        functionName: '',
        exportDefault: true,
        expected: '/foo',
      },
      {
        file: '/src/lambda/foo.ts',
        functionName: 'bar',
        exportDefault: false,
        expected: '/foo/bar',
      },
      {
        file: '/src/lambda/foo/bar/baz/qux.ts',
        functionName: 'getArticle',
        exportDefault: false,
        expected: '/foo/bar/baz/qux/getArticle',
      },
      {
        file: '/src/lambda/foo/baz.ts',
        functionName: 'bar',
        exportDefault: false,
        expected: '/foo/baz/bar',
      },
      {
        file: '/src/lambda/foo/[...baz].ts',
        functionName: 'bar',
        exportDefault: false,
        expected: '/foo/baz/bar/*',
      },
    ]

    for (const cse of cases) {
      expect(
        router.functionToHttpPath(cse.file, cse.functionName, cse.exportDefault)
      ).toEqual(cse.expected)
    }
  })

  it('file system route', () => {
    const cases = [
      // normal
      ['/', '/'],
      ['/api', '/api'],
      ['/api/foo', '/api/foo'],
      ['/api/foo/bar', '/api/foo/bar'],

      // index routes
      ['/index', '/'],
      ['/api/index', '/api'],
      ['/api/foo/index', '/api/foo'],

      // not index routes
      ['/index/index', '/index'],
      ['/api/index/foo', '/api/index/foo'],

      // dynamic params
      ['/[api]', '/:api'],
      ['/[api]/foo', '/:api/foo'],
      ['/api/[foo]/[bar]', '/api/:foo/:bar'],
      ['/api/foo/[bar]', '/api/foo/:bar'],

      // dynamic params + index routes
      ['/[index]', '/:index'],
      ['/[api]/index', '/:api'],
      ['/[api]/[index]/foo', '/:api/:index/foo'],
      ['/[api]/[index]/foo/index', '/:api/:index/foo'],

      // catchAll params
      ['/[...index]', '/*'],
      ['/[...api]', '/api/*'],
      ['/api/[...index]', '/api/*'],
      ['/api/foo/[bar]/[...baz]', '/api/foo/:bar/baz/*'],

      // + function name
      ['/', '/getArticle', 'getArticle'],
      ['/api', '/api/getArticle', 'getArticle'],

      ['/[api]', '/:api/getArticle', 'getArticle'],
      ['/[api]/foo', '/:api/foo/getArticle', 'getArticle'],

      ['/index', '/getArticle', 'getArticle'],
      ['/api/index', '/api/getArticle', 'getArticle'],
      ['/[index]', '/:index/getArticle', 'getArticle'],

      ['/[...index]', '/getArticle/*', 'getArticle'],
      ['/[...api]', '/api/getArticle/*', 'getArticle'],
      ['/api/[...index]', '/api/getArticle/*', 'getArticle'],
      [
        '/api/foo/[bar]/[...baz]',
        '/api/foo/:bar/baz/getArticle/*',
        'getArticle',
      ],
    ]

    for (const [input, expected, functionName] of cases) {
      expect(router.buildUrl(input, functionName)).toEqual(expected)
    }
  })

  it('buildUrl', () => {
    expect(() =>
      router.buildUrl('/[...index]/index.ts')
    ).toThrowErrorMatchingSnapshot()
  })
})
