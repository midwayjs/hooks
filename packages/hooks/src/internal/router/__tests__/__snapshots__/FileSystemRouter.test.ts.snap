// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewFileRouter buildUrl 1`] = `"Catch all routes must be the last part of the path. Current input: /[...index]/index.ts"`;

exports[`NewFileRouter test file router 1`] = `
Object {
  "baseDir": "lambda",
}
`;
